src/main/java/org/jenkinsci/plugins/workflow/log/DelayBufferedOutputStream.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2018 CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.workflow.log;
+
+import java.io.BufferedOutputStream;
+import java.io.FilterOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.lang.ref.Reference;
+import java.lang.ref.WeakReference;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import jenkins.util.Timer;
+
+/**
+ * Buffered output stream which is guaranteed to deliver content after some time even if idle and the buffer does not fill up.
+ * The automatic “flushing” does <em>not</em> flush the underlying stream, for example via {@code ProxyOutputStream.Flush}.
+ */
+final class DelayBufferedOutputStream extends BufferedOutputStream {
+
+    private static final Logger LOGGER = Logger.getLogger(DelayBufferedOutputStream.class.getName());
+
+    // TODO make these customizable (not trivial since this system properties would need to be loaded on the master side and then remoted)
+    private static final long MIN_RECURRENCE_PERIOD = 250; // ¼s
+    private static final long MAX_RECURRENCE_PERIOD = 10_000; // 10s
+    private static final float RECURRENCE_PERIOD_BACKOFF = 1.05f;
+
+    private long recurrencePeriod = MIN_RECURRENCE_PERIOD;
+
+    DelayBufferedOutputStream(OutputStream out) {
+        super(new FlushControlledOutputStream(out)); // default buffer size: 8Kib
+        reschedule();
+    }
+
+    private void reschedule() {
+        Timer.get().schedule(new Flush(this), recurrencePeriod, TimeUnit.MILLISECONDS);
+        recurrencePeriod = Math.min((long) (recurrencePeriod * RECURRENCE_PERIOD_BACKOFF), MAX_RECURRENCE_PERIOD);
```

Comment:
Why do we need an exponential backoff for simply flushing logs (which we're presumably doing to ensure we don't lose data in the event a master goes down)?  Wouldn't it be enough to just schedule this to run periodically at a pretty long interval (say 10-30s)?  Then it won't run that often, and a few seconds of log data isn't as big a deal.

Reply:
> which we're presumably doing to ensure we don't lose data in the event a master goes down

No, that is a consideration, but not the driver for the choice of `MIN_RECURRENCE_PERIOD`. Consider a user watching a build that is doing different stuff, including a number of quick `sh` steps. You want to see output from those in as close to real time as possible. But then once a given process has been running for a few minutes, the likelihood that it is being watched closely drops off and it is fine to send big chunks at intervals.

Reply:
FWIW: I had a similar question, see https://github.com/jenkinsci/workflow-api-plugin/pull/81#discussion_r228180118 for related discussion.